### ISSUE TITLE
Allow processing user attributes before they are stored

### ISSUE THREAD

**Issue by pedroigor** [keycloak/keycloak#23222]
### Description

The [User Profile Design Document](https://github.com/keycloak/keycloak-community/blob/main/design/user-profile.md#converter-optional) defines a `converter` as a mechanism to process raw attributes before they are validated or set as a user attribute.

I'm not sure if the name `converter` is the best because most of the time we are going to transform values from one format to another. Perhaps `transformers` is a more suitable and generic name.

There are different use cases where converters might be useful:

* https://github.com/keycloak/keycloak/issues/11757
* https://github.com/keycloak/keycloak/issues/11452

A converter should work similarly to how we configure/manage validators on a per-attribute basis. Based on that, we can introduce a property `converters` to the attribute configuration similar to the following:

```
{
	"name": "username",
	"converters": {
		"converter-id": { "config-key": "config-value"}
                ...
	}
}
```

# Converter SPI
From a Java API perspective, we can introduce a new SPI. Also similar to how we are using the Validation SPI .

A converter could look like:

```
public interface Converter extends Provider {

    String convert(AttributeContext context);
}
```

The method `Converter.convert` should be invoked before the validation phase so that validators can run using the expected value. Once the passing the validation phase, the value is stored

Different from the `Validation SPI` we could obtain the converter config from the `AttributeContext`. 

Another approach is to leverage the API from `ValidatorConfig` to build the `ConverterConfig` by extracting a more generic configuration type from `ValidatorConfig`. And then have the `ConverterConfig` as a parameter.

## Alternatives to building our own

I'm not sure about introducing and leveraging additional dependencies or relying on existing ones that provide a similar API and level of configuration.

# UI
We need discussions with the @keycloak/ui-maintainers to see how it would best fit into the administration console.

### Motivation

Allow processing user attributes before they are stored

**Comment by sschu** [comment]
I would also prefer `transformer`, it is also in line with usage in e.g. ETL.

**Comment by jbman** [comment]
Not sure about "convert should be invoked before the validation phase". A validated value could also be transformed after validation, e.g truncated that it fits into storage or enhanced with additional information. But this would only work if validators SPI is enhanced so that implementations can  transform the value and validators would need an order.

**Comment by pedroigor** [comment]
I see. Could you elaborate on the use case where you want to transform the value after validation and not the opposite?

The motivation for having it running before the validation phase is that validators operate on a specific format/type/pattern. Transformation is the attempt to prepare the value for the next phase: validation.

**Comment by pedroigor** [comment]
I would start this task by running transformers first and without expecting any validation to happen from there. The two use cases I linked here (the issues in the description) are about transforming a value and preparing it for the next phase.

I see the point from @jbman where you might want to "truncate to fit into storage or enhance it", but I think truncating a value is not really a user profile concern (but storage). We also have the `AttributeChangeListener` that can be used to perform additional steps when storing attributes.

**Comment by jbman** [comment]
I just remembered that a customer wanted to map IdP claims into to a user attribute. If values are to long, they can't be stored. A validator could validate a an URI and a transformer could truncate it, which may be better than discarding the value completely.

Overall I think a design with ordered validators and transforms leads to more fine grained reusable validators and transformers. "Transform then validate" leads to transformers which will sometimes validate to their transformation.
Regarding the date use case: to transform a date from one syntax  A into another B, it must be validated. The sequence would be validate format A, transform A to B. 
Regarding the trim use case: Trimming will always go first.

That said, transformers first will work and bring the options to solve the requested cases.

**Comment by ghenadiibatalski** [comment]
@pedroigor Looks to me like an interesting and useful feature (e.g., converting an ssh-rsa key into RFC4716 format). Are there any plans for when it could be available for preview?
Regards, Gena

**Comment by pedroigor** [comment]
@ghenadiibatalski Yeah, I think we need to prioritize this one now the declarative user profile is supported/default.

We have other things in our plate right now but I'll mark this one as important so we can consider introducing it ASAP.
